src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/Windows/PkcsPalWindows.Encrypt.cs

```diff
@@ -0,0 +1,364 @@
+
+                    case AlgId.CALG_DH_SF:
+                    case AlgId.CALG_DH_EPHEM:
+                        recipientChoice = CMsgCmsRecipientChoice.CMSG_KEY_AGREE_RECIPIENT;
+                        break;
+
+                    default:
+                        throw ErrorCode.CRYPT_E_UNKNOWN_ALGO.ToCryptographicException();
+                }
+
+                CMSG_RECIPIENT_ENCODE_INFO recipientEncodeInfo;
+                recipientEncodeInfo.dwRecipientChoice = recipientChoice;
+
+                unsafe
+                {
+                    switch (recipientChoice)
+                    {
+                        case CMsgCmsRecipientChoice.CMSG_KEY_TRANS_RECIPIENT:
+                            recipientEncodeInfo.pCmsRecipientEncodeInfo = (IntPtr)EncodeKeyTransRecipientInfo(recipient, hb);
+                            break;
+
+                        case CMsgCmsRecipientChoice.CMSG_KEY_AGREE_RECIPIENT:
+                            recipientEncodeInfo.pCmsRecipientEncodeInfo = (IntPtr)EncodeKeyAgreeRecipientInfo(recipient, contentEncryptionAlgorithm, hb);
+                            break;
+
+                        default:
+                            throw ErrorCode.CRYPT_E_UNKNOWN_ALGO.ToCryptographicException();
+                    }
+                }
+                return recipientEncodeInfo;
+            }
+
+            //
+            // This returns an allocated native memory block. Its lifetime (and that of any allocated subblocks it may point to) is that of "hb". 
+            //
+            private static unsafe CMSG_KEY_TRANS_RECIPIENT_ENCODE_INFO* EncodeKeyTransRecipientInfo(CmsRecipient recipient, HeapBlockRetainer hb)
+            {
+                // "recipient" is a deep-cloned CmsRecipient object whose lifetime this class controls. Because of this, we can pull out the CERT_CONTEXT* and CERT_INFO* pointers without
+                // bringing in all the SafeCertContextHandle machinery, and embed pointers to them in the memory block we return. Yes, this code is scary.
+                using (SafeCertContextHandle hCertContext = recipient.Certificate.CreateCertContextHandle())
+                {
+                    CERT_CONTEXT* pCertContext = hCertContext.DangerousGetCertContext();
+                    CERT_INFO* pCertInfo = pCertContext->pCertInfo;
+
+                    CMSG_KEY_TRANS_RECIPIENT_ENCODE_INFO* pEncodeInfo = (CMSG_KEY_TRANS_RECIPIENT_ENCODE_INFO*)(hb.Alloc(sizeof(CMSG_KEY_TRANS_RECIPIENT_ENCODE_INFO)));
+
+                    pEncodeInfo->cbSize = sizeof(CMSG_KEY_TRANS_RECIPIENT_ENCODE_INFO);
+
+                    CRYPT_ALGORITHM_IDENTIFIER algId = pCertInfo->SubjectPublicKeyInfo.Algorithm;
+                    pEncodeInfo->KeyEncryptionAlgorithm = algId;
+
+                    pEncodeInfo->pvKeyEncryptionAuxInfo = IntPtr.Zero;
+                    pEncodeInfo->hCryptProv = IntPtr.Zero;
+
+                    pEncodeInfo->RecipientPublicKey = pCertInfo->SubjectPublicKeyInfo.PublicKey;
```

Comment:
Ah, right, okay.  Since CertDuplicateContextHandle is documented as AddRef (instead of clone) I agree that we're safe here.